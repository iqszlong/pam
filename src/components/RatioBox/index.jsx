import React from 'react';
import Style from './style.css';

class RatioBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
        ratio:(9/16)*100,
    };
  }

  static getDerivedStateFromProps(props, state) {
      const data = state;
    //渲染之前修改state值
    if(props.width !== undefined && props.height !== undefined){
        data.ratio = (props.height/props.width)*100
        return data;
    }
    // console.info(props, state);
    return null;
  }

  shouldComponentUpdate(newProps, newState) {
    // console.log('shouldComponentUpdate');
    //提高性能 默认返回true this.state !== newState
    
    return this.props !== newProps;
  }


  render() {
    return (
      <>
        <div className={`${Style.layout} ${this.props.className || ''}`} style={{paddingTop:`${this.state.ratio}%`}}>
          <div className={Style.inner}>{this.props.children}</div>
        </div>
      </>
    );
  }

  componentDidMount() {
    //挂载完成之后，请求数据更新state
    // console.log(this.props);

  }
}

export default RatioBox;
