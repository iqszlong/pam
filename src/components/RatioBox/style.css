.layout{
    position: relative;
    display: block;
    height: 0;
    padding: 0;
    overflow: hidden; 
    background-color: var(--embed-bg,#eee);
}
.inner{
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border: 0;
}