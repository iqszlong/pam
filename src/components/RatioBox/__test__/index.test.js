import RatioBox from '..';
import renderer from 'react-test-renderer';

import { shallow ,render,mount} from 'enzyme';



describe('html测试', () => {
    it('标签测试', () => {
        const wrapper = renderer.create(<RatioBox/>);
        expect(wrapper.root.children.length).toBe(1);
        const outerLayer = wrapper.root.children[0];
        expect(outerLayer.type).toBe('div');
    });
    test('渲染测试', () => {
        const wrapper = mount(<RatioBox></RatioBox>);
        expect(wrapper).toMatchSnapshot();
    });
});

describe('属性测试', () => {
    test('className', () => {
        const wrapper = shallow(<RatioBox className="test_css" width="10" height="5" />);
        // console.log(wrapper.debug());
        const outerLayer = wrapper.find('.layout');
        // console.log(outerLayer.debug());
        expect(outerLayer.prop('className')).toMatch('test_css');
        expect(outerLayer.prop('style')).toEqual({paddingTop:"50%"});
    });
});