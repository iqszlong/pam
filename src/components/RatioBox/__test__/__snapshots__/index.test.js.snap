// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html测试 渲染测试 1`] = `ReactWrapper {}`;
